app/helpers/heatmap_helper.rb

```diff
@@ -71,25 +35,434 @@ def sample_taxons_dict(params)
+                        .each { |row| row[:filtered] = HeatmapHelper.check_custom_filters(row, threshold_filters) }
+
+        results[sample_id] = {
+          sample_id: sample_id,
+          name: samples_by_id[sample_id].name,
+          metadata: samples_by_id[sample_id].metadata_with_base_type,
+          host_genome_name: samples_by_id[sample_id].host_genome_name,
+          taxons: filtered_rows
+        }
+      end
+    end
+
+    # For samples that didn't have matching taxons, just throw in the metadata.
+    samples.each do |sample|
+      unless results.key?(sample.id)
+        results[sample.id] = {
+          sample_id: sample.id,
+          name: sample.name,
+          metadata: sample.metadata_with_base_type,
+          host_genome_name: sample.host_genome_name
+        }
+      end
+    end
+
+    # Flatten the hash
+    results.values
+  end
+
+  def self.fetch_samples_taxons_counts(samples, taxon_ids, parent_ids, background_id)
+    parent_ids = parent_ids.to_a
+    parent_ids_clause = parent_ids.empty? ? "" : " OR taxon_counts.tax_id in (#{parent_ids.join(',')}) "
+
+    # Note: subsample_fraction is of type 'float' so adjusted_total_reads is too
+    # Note: stdev is never 0
+    # Note: connection.select_all is TWICE faster than TaxonCount.select
+    # (I/O latency goes from 2 seconds -> 0.8 seconds)
+    # Had to derive rpm and zscore for each sample
+    sql_results = TaxonCount.connection.select_all("
+      SELECT
+        taxon_counts.pipeline_run_id     AS  pipeline_run_id,
+        taxon_counts.tax_id              AS  tax_id,
+        taxon_counts.count_type          AS  count_type,
+        taxon_counts.tax_level           AS  tax_level,
+        taxon_counts.genus_taxid         AS  genus_taxid,
+        taxon_counts.family_taxid        AS  family_taxid,
+        taxon_counts.name                AS  name,
+        taxon_counts.superkingdom_taxid  AS  superkingdom_taxid,
+        taxon_counts.is_phage            AS  is_phage,
+        taxon_counts.count               AS  r,
+        taxon_summaries.stdev            AS stdev,
+        taxon_summaries.mean             AS mean,
+        taxon_counts.percent_identity    AS  percentidentity,
+        taxon_counts.alignment_length    AS  alignmentlength,
+        IF(
+          taxon_counts.e_value IS NOT NULL,
+          (0.0 - taxon_counts.e_value),
+          #{ReportHelper::DEFAULT_SAMPLE_NEGLOGEVALUE}
+        )                                AS  neglogevalue,
+        taxon_counts.percent_concordant  AS  percentconcordant
+      FROM taxon_counts
+      LEFT OUTER JOIN taxon_summaries ON
+        #{background_id.to_i}   = taxon_summaries.background_id   AND
+        taxon_counts.count_type = taxon_summaries.count_type      AND
+        taxon_counts.tax_level  = taxon_summaries.tax_level       AND
+        taxon_counts.tax_id     = taxon_summaries.tax_id
+      WHERE
+        pipeline_run_id IN (
+          SELECT MAX(id)
```

Comment:
Max id is not necessarily the last run... This is something we need to fix in the entire app...eventually.

Reply:
good point ... should disucss in person